# Part B
#### Grading Criteria
Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [1]:
# perform installs
!pip install numpy==2.0.2
!pip install pandas==2.2.2
!pip install tensorflow_cpu==2.18.0
!pip install scikit-learn

In [2]:
# imports
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input

import warnings
warnings.simplefilter('ignore', FutureWarning)

2025-03-02 05:15:29.226296: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-03-02 05:15:29.285517: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [3]:
# Get concrete dataset
filepath='https://cocl.us/concrete_data'
concrete_data = pd.read_csv(filepath)

concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
# check how many data points we have
concrete_data.shape

(1030, 9)

In [5]:
# look at dataset info
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
# make sure we don't have null values
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [7]:
# Split dataset into predictors (all columns except Strength) and target (Strength column)
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [8]:
# normalize predictors
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [9]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

In [10]:
# function to build model
def build_model_part_b():
    # create model
    model = Sequential()
    model.add(Input(shape=(n_cols,)))
    # model has hidden layer with 10 nodes and relu act func
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error') # Use adam and mse
    return model

In [11]:
# build model
part_b_model = build_model_part_b()

In [12]:
mse_list = []

In [13]:
def test_pt_b_mse():
    # split into train and test sets and hold 30% of data for testing
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
    # train model with 50 epochs
    part_b_model.fit(X_train, y_train, epochs=50, verbose=2)
    y_pred = part_b_model.predict(X_test)  #eval model on test data
    mse = mean_squared_error(y_test, y_pred)  # compute mse
    mse_list.append(mse)  #add to list of mses

In [14]:
# repeat 50 times
num_iterations = 50
for i in range(num_iterations):
    test_pt_b_mse()

Epoch 1/50
23/23 - 1s - 33ms/step - loss: 1559.9924
Epoch 2/50
23/23 - 0s - 5ms/step - loss: 1543.2361
Epoch 3/50
23/23 - 0s - 4ms/step - loss: 1526.3136
Epoch 4/50
23/23 - 0s - 5ms/step - loss: 1509.0848
Epoch 5/50
23/23 - 0s - 4ms/step - loss: 1491.2827
Epoch 6/50
23/23 - 0s - 3ms/step - loss: 1472.7035
Epoch 7/50
23/23 - 0s - 4ms/step - loss: 1453.4072
Epoch 8/50
23/23 - 0s - 3ms/step - loss: 1432.8838
Epoch 9/50
23/23 - 0s - 5ms/step - loss: 1411.1720
Epoch 10/50
23/23 - 0s - 4ms/step - loss: 1388.2388
Epoch 11/50
23/23 - 0s - 4ms/step - loss: 1363.4209
Epoch 12/50
23/23 - 0s - 4ms/step - loss: 1337.3274
Epoch 13/50
23/23 - 0s - 4ms/step - loss: 1309.5038
Epoch 14/50
23/23 - 0s - 4ms/step - loss: 1280.6174
Epoch 15/50
23/23 - 0s - 4ms/step - loss: 1250.1735
Epoch 16/50
23/23 - 0s - 4ms/step - loss: 1218.3463
Epoch 17/50
23/23 - 0s - 4ms/step - loss: 1185.9951
Epoch 18/50
23/23 - 0s - 4ms/step - loss: 1152.3721
Epoch 19/50
23/23 - 0s - 4ms/step - loss: 1118.3500
Epoch 20/50
23/23 - 

In [15]:
# Report the mean and the standard deviation of the mean squared errors.
average_mse = np.mean(mse_list)
print(f"Average Mean Squared Error over {num_iterations} iterations: {average_mse}")
std_mse = np.std(mse_list)
print(f"Mean Squared Error Standard Deviation over {num_iterations} iterations: {std_mse}")

Average Mean Squared Error over 50 iterations: 51.930608871813966
Mean Squared Error Standard Deviation over 50 iterations: 36.85541987092113


In [1]:
# Part A metrics:
# Average Mean Squared Error over 50 iterations: 66.2795128282057
# Mean Squared Error Standard Deviation over 50 iterations: 60.89369286469975

# How does the mean of the mean squared errors compare to that from Step A?
For part A, we had a mean squared error of 66.2795128282057 and a standard deviation of 60.89369286469975.
For part B, we had a mean squared error of 51.930608871813966 and a standard deviation of 36.85541987092113.
Based on the lower MSE and lower standard deviation, we can conclude that normalization of data significantly improves the model's performance on this dataset.